### PR TITLE
feat(useListNavigation, Composite): support grid navigation over items with variable sizes

### PR DESCRIPTION
**Pull request by JavierM42**
`useListNavigation` and `Composite` both assume that grids will be uniform (all items span 1 row and 1 column). We came across the need for items with variable sizes and decided to implement it.

If it's a welcome feature, we'd be happy to see it added to the library.

With the proposed changes, `useListNavigation` and `Composite` accept would two new optional props:
  - `itemSizes`: an array of `{ width: number; height: number }`, which represent the row and column spans of each item (defaults to all items being 1x1).
  - `dense`: a boolean flag indicating if the grid follows [CSS grid-auto-flow](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-auto-flow) `dense` placement rules or not.
  
The navigation logic is quite complex but I hope it's commented and tested with enough clarity.
  
## Demo video:

https://github.com/floating-ui/floating-ui/assets/32175149/e8b356e2-1ffe-4ced-88f3-7afedb0c77c1

---

Huge thanks to @dantevvp who originally needed this feature and then helped so much to implement it.

### CODE REVIEW

File: packages/react/src/components/Composite.tsx [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import type {Size} from '@floating-ui/utils';
```
Comment (atomiks) [inline packages/react/src/components/Composite.tsx:1]: `Dimensions` type should be the same?

File: packages/react/src/components/Composite.tsx [hunk 2]
```diff
@@ -88,25 +104,70 @@ export const Composite = React.forwardRef<
...
+      const sizes = itemSizes ?? Array.from(
+        Array(elementsRef.current.length),
+        () => ({ width: 1, height: 1 })
+      );
+      // To calculate movements on the grid, we use hypothetical cell indices
+      // as if every item was 1x1, then convert back to real indices.
+      const cellMap = buildCellMap(sizes, cols, dense);
+      const minGridIndex = cellMap.findIndex(
+        index => index !== undefined && !disabledIndices.includes(index)
+      );
```
Comment (atomiks) [inline packages/react/src/components/Composite.tsx:119]: `index != null` is a bit shorter

File: packages/react/src/utils/composite.ts [hunk 3]
```diff
@@ -233,3 +234,72 @@ export function getGridNavigatedIndex(
...
+
+/** For each cell index, gets the item index that occupies that cell */
+export function buildCellMap(sizes: Size[], cols: number, dense: boolean) {
+  const cellMap: (number | undefined)[] = [];
+  let startIndex = 0;
+  sizes.forEach(({ width, height }, index) => {
+    if (width > cols) {
+      throw new Error(
+        `Invalid grid: item width at index ${index} is greater than grid columns`
+      );
```
Comment (atomiks) [inline packages/react/src/utils/composite.ts:246]: Wrap in `if (__DEV__)`

File: packages/utils/src/types.ts [hunk 4]
```diff
@@ -6,7 +6,9 @@ export type Strategy = 'absolute' | 'fixed';
 export type Axis = 'x' | 'y';
 export type Length = 'width' | 'height';
 export type SideObject = {[key in Side]: number};
-export type Rect = {x: number; y: number; width: number; height: number};
+export type Position = {[key in Axis]: number}
+export type Size = {[key in Length]: number};
```
Comment (atomiks) [inline packages/utils/src/types.ts:10]: `Coords` and `Dimensions` are these types currently

File: packages/react/src/utils/composite.ts [hunk 5]
```diff
@@ -241,7 +241,7 @@ export function buildCellMap(sizes: Dimensions[], cols: number, dense: boolean)
   const cellMap: (number | undefined)[] = [];
   let startIndex = 0;
   sizes.forEach(({ width, height }, index) => {
-    if (width > cols) {
+    if (width > cols && __DEV__) {
       throw new Error(
         `Invalid grid: item width at index ${index} is greater than grid columns`
       );
```
Comment (atomiks) [inline packages/react/src/utils/composite.ts]: Some bundlers don't like it not to be on its own:

```js
if (width > cols) {
  if (__DEV__) {
    throw...
  }
}
```

**Comment by atomiks** [comment]
This is cool thanks 👍  when testing out `ComplexGrid` locally, the styles don't seem to show variable sizes. Is it possible to:

1. Add a route `complex-grid` (I added it manually)
2. Edit the item styles to show disabled styles + variable sizes?

**Comment by atomiks** [comment]
Seems good to me! Type conflict appeared due to the PR that just got merged

**Comment by JavierM42** [comment]
Rebased

**Comment by atomiks** [comment]
Sorry, something went wrong again :\ will try to merge ASAP before any new conflicts

**Comment by JavierM42** [comment]
No worries!

Rebased.

**Comment by JavierM42** [comment]
No worries! Conflicts fixed.

**Comment by atomiks** [comment]
@JavierM42 it seems there's an issue with the Emoji Picker demo with this PR. Not sure why, but it seems whatever logic it has isn't being tested so regression wasn't caught 🤔

**Comment by JavierM42** [comment]
@atomiks I opened https://github.com/floating-ui/floating-ui/pull/2696, but I have more questions than answers 😅

**Pull request merged**